---
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
output: 
  html_document
params:
  master_db:
    value: x
  patient_vcf:
    value: x
  failed:
    value: x
  patient_ID:
    value: x
---
```{r, include = FALSE}
knitr::opts_chunk$set(echo = F,
                      warning = F,
                      message = F)
```

```{r, include = FALSE}
library(rmarkdown)
library(data.table)
library(dplyr)
library(tidyr)
#gemini_db <- 'test_40.db'
#file <- '17-1012_RD_v11_assay.VEP.GRCh37.anno.vcf.gz'
#setwd('/data/mcgaugheyd/projects/nei/hufnagel/casey_panels/test2')

# read in vcf to count number of header rows to skip
file = params$patient_vcf
con <- file(file, open = 'r')
data <- readLines(con)
row_skip <- sum(sapply(data, function(x) ifelse(substr(x,1,2)=='##', 1,0)))
# now use fread to pull in vcf, skipping header rows
vcf <- fread(paste('zcat', file), skip=row_skip)
id_status <- vcf %>% dplyr::select(`#CHROM`, POS, ID, REF, ALT, SAMPLE) %>% tidyr::separate(SAMPLE, c('GT', 'CT', 'JC'), ':')

primary_IDs <- id_status %>% dplyr::filter(JC=='Primary') %>% .[['ID']]
secondary_IDs <- id_status %>% dplyr::filter(JC=='Secondary') %>% .[['ID']]

gemini_query <- function(gemini_db, vcf_id, sample_name, ref, alt){
  # Queries gemini in two ways:
  ## With the HGVS as the search term
  ## If ref, alt given, then uses HGVS as a fuzzy match and ref alt to confirm
  base_query <- "gemini query --show-samples --header -q 'select chrom, start, end, ref, alt, vcf_id, hgvsc, hgvsp, type, gene, impact, rs_ids, pfam_domain, clinvar_diseases, clinvar_ID, clinvar_pathogenic, clinvar_sig, hgmd_overlap, af_exac_afr, af_exac_all af_exac_amr, af_exac_eas, af_exac_nfe af_exac_oth, af_exac_sas, an_exac_all, exac_num_het, exac_num_hom_alt, mis_z, pli, gerp_elements, polyphen_pred, polyphen_score, sift_pred, sift_score, cadd_phred FROM variants "
  base_query <- gsub(' FROM', paste0(', gts.', sample_name, ' FROM'), base_query)
  if (missing(ref) & missing(alt)){
    out <- system(paste0(base_query,
                         "WHERE vcf_id IN (", 
                         paste0('\"',vcf_id, '\"', collapse=','), ") ' ", 
                         gemini_db), 
                  intern=TRUE) 
  }
  else {
    out <- system(paste0(base_query,
                         "WHERE vcf_id LIKE ", 
                         paste0('\"%',vcf_id, '%\"', collapse=','), 
                         " AND ref IS \"", ref, 
                         "\" AND alt IS \"", alt,"\"' ", 
                         gemini_db), 
                  intern=TRUE) 
  }
  return(out)
}

DF_maker <- function(df_vector, priority){
  # Converts character vector of gemini captured call into a data frame
  DF <- data.frame(do.call(rbind, strsplit(df_vector, '\t')))
  colnames(DF) <- as.character(unlist(DF[1,]))
  DF <- DF[-1,] 
  DF$Priority <- priority
  return(DF)
}
ID_to_DF <- function(IDs, priority){
  # Converts ID (HGVS) vector to data frame with status 
  id_DF <- data.frame(IDs)
  id_DF$Priority <- priority
  colnames(id_DF)[1] <- 'HGVS'
  return(id_DF)
}
```

```{r, include = FALSE}
gemini_db <- params$master_db
sample_name <- params$patient_ID
print(sample_name)
old_sample <- sample_name
sample_name <- gsub('-','_', sample_name)
if (length(primary_IDs>0)){
  primary_DF <- DF_maker(gemini_query(gemini_db, primary_IDs, sample_name), 'Primary')
  primary_IDs <- ID_to_DF(primary_IDs, 'Primary')
} else {
  primary_DF <- NA
  primary_IDs <- NA
}
if (length(secondary_IDs)>0){
  secondary_DF <- DF_maker(gemini_query(gemini_db, secondary_IDs, sample_name), 'Secondary')
  secondary_IDs <- ID_to_DF(primary_IDs, 'Secondary')
} else {
  secondary_DF <- NA
  secondary_IDs <- NA
}

all_DF <- rbind(primary_DF, secondary_DF)
all_IDs <- rbind(primary_IDs, secondary_IDs)

#find failures to match ID/vcf_id and add
HGVS = all_IDs$HGVS
if (length(HGVS[!HGVS %in% all_DF$vcf_id]) > 0){
  for (i in HGVS[!HGVS %in% all_DF$vcf_id]){
    ref <- id_status %>% filter(ID==i) %>% .[['REF']]
    alt <- id_status %>% filter(ID==i) %>% .[['ALT']]
    query_result <- gemini_query(gemini_db, i, sample_name, ref, alt)
    if (length(query_result) != 2) {
      print(paste("ERROR: missing ", i)) 
      quit(status = 1)
    }
    priority <- all_IDs %>% filter(HGVS==i) %>% pull(Priority)
    all_DF <- rbind(all_DF, DF_maker(gemini_query(gemini_db, i, sample_name, ref, alt), priority) )
  }
}
```

```{r, include = F}
# Helper scripts to generate URLs for the tables
link_generator <- function(base_url, ID){
  if (is.na(ID) | is.na(base_url) | ID == 'None' | base_url == 'None'){
    url <- NA
  }
  else {
    ends <- strsplit(ID, ',')[[1]]
    url <- ''
    for (end in ends){
      url <- paste(paste0('<a href="', base_url, end, '">', end, '</a> '), url)
    }
    if (ID == 'None') {
      url <- NA
    }
  }
  return(as.character(url))
}
clinvar_link_generator <- function(base_url, ID, clinvar_sig){
  if (clinvar_sig == 'None' | is.na(clinvar_sig) | is.na(ID) | ID == 'None') {
    url <- NA
  }
  else {
    ends <- strsplit(ID, ',')[[1]]
    url <- ''
    for (end in ends){
      url <- paste(paste0('<a href="', base_url, end, '">', as.character(clinvar_sig), '</a> '), url)
    }
  }
  return(as.character(url))
}
google_link_generator <- function(var, gene){
  if (is.na(var) || is.na(gene) || var == 'None' || gene == 'None'){
    output <- NA
  }
  else {
    url = 'https://scholar.google.com/scholar?hl=en&q='
    var <- gsub('>','%3E',var)
    var <- strsplit(var, ':')[[1]][2]
    output <- paste0('<a href="', url, var, '+', gene,'">', 'Feeling&nbspLucky?', '</a>')
  }
  return(output)
}
```

```{r, include = FALSE}
failed <- fread(params$failed, header = F)
colnames(failed) <- c('vcf_id','Priority')
```


```{r, include=F}
# prep for printing
all_DF_view <- all_DF %>% mutate(ID=paste(chrom,end,ref,alt,sep='-')) %>% select(ID, vcf_id:Priority) %>% bind_rows(failed) %>% mutate_if(is.factor, as.character)
# find gt index
gts_index <- grep('gts',colnames(all_DF_view))
colnames(all_DF_view)[gts_index] <- 'Genotype'
  
# select indices for different column views
all_cols <- seq(1,ncol(all_DF_view))
exac <- c(2, gts_index, 6,grep('exac', colnames(all_DF_view), ignore.case = T))
neg_exac <- setdiff(all_cols, exac) - 1
core <- c(1, gts_index, 2,4,6,7,8,13)
neg_core <- setdiff(all_cols, core) - 1

# format DF
all_DF_view <- all_DF_view %>% 
    dplyr::filter(Priority == 'Primary') %>% 
    dplyr::mutate(clinvar_diseases = gsub('\\|',' ', as.character(clinvar_diseases)),
                  variant_samples = gsub(',', ', ', as.character(variant_samples)),
                  het_samples = gsub(',', ', ', as.character(het_samples)),
                  hom_alt_samples = gsub(',', ', ', as.character(hom_alt_samples))) %>% 
    dplyr::rowwise() %>% 
    dplyr::mutate(hgvsp = strsplit(as.character(hgvsp),':')[[1]][2],
                  ID=link_generator('http://gnomad.broadinstitute.org/variant/',as.character(ID)),
                  clinvar_sig = clinvar_link_generator('https://www.ncbi.nlm.nih.gov/clinvar?term=',as.character(rs_ids), as.character(clinvar_sig)),
                  rs_ids=link_generator('https://www.ncbi.nlm.nih.gov/projects/SNP/snp_ref.cgi?rs=',as.character(rs_ids)),
                  `Google Scholar` = google_link_generator(as.character(vcf_id), as.character((gene))),
                  vcf_id = link_generator('https://www.mutalyzer.nl/name-checker?description=',as.character(vcf_id))) %>% data.table()
```
## `r gsub('_',' ', old_sample)` {.tabset .tabset-fade}
### Primary Variants 
```{r}
if (nrow(all_DF_view %>% filter(Priority == 'Primary')) == 0){
  out <- 'No primary variants'
} else {
  out <- all_DF_view %>% 
    dplyr::filter(Priority == 'Primary') %>% 
    DT::datatable(width=1200, 
                  escape = FALSE,
                  rownames = F, 
                  class='compact', 
                  extensions = c('FixedColumns','Buttons'), 
                  options = list(columnDefs=list(list(targets=neg_core, visible = FALSE)),
                                 pageLength = 30,
                                 dom = 'Bfrtip',
                                 buttons = list(list(extend='pageLength'),
                                                list(extend='colvis', collectionLayout='four-column'),
                                                list(extend='colvisGroup', text='Core Fields', hide=paste(neg_core, collapse = ','), show = paste(core, collapse = ',')),
                                                list(extend='colvisGroup', text='ExAC', hide=paste(neg_exac,collapse = ','), show = paste(exac, collapse = ',' )),
                                                list(extend='colvisGroup', text='Show All', show = ':hidden' )),
                                 columnDefs=list(list(targets=c(1), class="dt-left")))) %>% 
    DT::formatSignif(columns=c('af_exac_afr','af_exac_amr','af_exac_eas','af_exac_oth','af_exac_sas','an_exac_all','mis_z','pli'),3)
}
out
```

### Secondary Variants
```{r}
if (nrow(all_DF_view %>% filter(Priority == 'Secondary')) == 0){
  out <- 'No secondary variants'
} else {
  out <- all_DF_view %>% 
    dplyr::filter(Priority == 'Secondary') %>% 
    DT::datatable(width=1200, 
                  escape = FALSE,
                  rownames = F, 
                  class='compact', 
                  extensions = c('FixedColumns','Buttons'), 
                  options = list(columnDefs=list(list(targets=neg_core, visible = FALSE)),
                                 pageLength = 30,
                                 dom = 'Bfrtip',
                                 buttons = list(list(extend='pageLength'),
                                                list(extend='colvis', collectionLayout='four-column'),
                                                list(extend='colvisGroup', text='Core Fields', hide=paste(neg_core, collapse = ','), show = paste(core, collapse = ',')),
                                                list(extend='colvisGroup', text='ExAC', hide=paste(neg_exac,collapse = ','), show = paste(exac, collapse = ',' )),
                                                list(extend='colvisGroup', text='Show All', show = ':hidden' )),
                                 columnDefs=list(list(targets=c(1), class="dt-left")))) %>% 
    DT::formatSignif(columns=c('af_exac_afr','af_exac_amr','af_exac_eas','af_exac_oth','af_exac_sas','an_exac_all','mis_z','pli'),3)
}
out
```
