---
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
output: 
  html_document:
    css: datatable_fix.css
params:
  master_db:
    value: x
  patient_vcf:
    value: x
  failed:
    value: x
  patient_ID:
    value: x
---
```{r, include = FALSE}
knitr::opts_chunk$set(echo = F,
                      warning = F,
                      message = F)
```

```{r, include = FALSE}
library(rmarkdown)
library(data.table)
library(dplyr)
library(tidyr)
#gemini_db <- 'test_40.db'
#file <- '17-1012_RD_v11_assay.VEP.GRCh37.anno.vcf.gz'
#setwd('/data/mcgaugheyd/projects/nei/hufnagel/casey_panels/test2')

# read in vcf to count number of header rows to skip
file = params$patient_vcf
con <- file(file, open = 'r')
data <- readLines(con)
row_skip <- sum(sapply(data, function(x) ifelse(substr(x,1,2)=='##', 1,0)))
# now use fread to pull in vcf, skipping header rows
vcf <- fread(paste('zcat', file), skip=row_skip)
id_status <- vcf %>% dplyr::select(`#CHROM`, POS, ID, REF, ALT, SAMPLE) %>% tidyr::separate(SAMPLE, c('GT', 'CT', 'JC'), ':')

primary_IDs <- id_status %>% dplyr::filter(JC=='Primary') %>% .[['ID']]
secondary_IDs <- id_status %>% dplyr::filter(JC=='Secondary') %>% .[['ID']]

gemini_query <- function(gemini_db, vcf_id, sample_name, ref, alt){
  # Queries gemini in two ways:
  ## With the HGVS as the search term
  ## If ref, alt given, then uses HGVS as a fuzzy match and ref alt to confirm
  base_query <- "gemini query --show-samples --header -q 'select chrom, start, end, ref, alt, vcf_id, hgvsc, hgvsp, type, gene, impact, impact_severity, rs_ids, pfam_domain, clinvar_diseases, clinvar_ID, clinvar_pathogenic, clinvar_sig, hgmd_overlap, num_het, num_hom_alt, max_aaf_all, af_exac_afr, af_exac_all af_exac_amr, af_exac_eas, af_exac_nfe af_exac_oth, af_exac_sas, an_exac_all, exac_num_het, exac_num_hom_alt, mis_z, pli, gerp_elements, polyphen_pred, polyphen_score, sift_pred, sift_score, cadd_phred FROM variants "
  base_query <- gsub(' FROM', paste0(', gts.', sample_name, ' FROM'), base_query)
  if (missing(ref) & missing(alt)){
    out <- system(paste0(base_query,
                         "WHERE vcf_id IN (", 
                         paste0('\"',vcf_id, '\"', collapse=','), ") ' ", 
                         gemini_db), 
                  intern=TRUE) 
  }
  else {
    out <- system(paste0(base_query,
                         "WHERE vcf_id LIKE ", 
                         paste0('\"%',vcf_id, '%\"', collapse=','), 
                         " AND ref IS \"", ref, 
                         "\" AND alt IS \"", alt,"\"' ", 
                         gemini_db), 
                  intern=TRUE) 
  }
  return(out)
}

DF_maker <- function(df_vector, priority){
  # Converts character vector of gemini captured call into a data frame
  DF <- data.frame(do.call(rbind, strsplit(df_vector, '\t')))
  colnames(DF) <- as.character(unlist(DF[1,]))
  DF <- DF[-1,] 
  DF$Priority <- priority
  return(DF)
}
ID_to_DF <- function(IDs, priority){
  # Converts ID (HGVS) vector to data frame with status 
  id_DF <- data.frame(IDs)
  id_DF$Priority <- priority
  colnames(id_DF)[1] <- 'HGVS'
  return(id_DF)
}
```

```{r, include = FALSE}
# pull parameters for gemini db and sample name
gemini_db <- params$master_db
sample_name <- params$patient_ID
print(sample_name)
old_sample <- sample_name
# sql doesn't accept '-'. Gemini autochanges these to _
sample_name <- gsub('-','_', sample_name)
# query gemini, save to df
if (length(primary_IDs>0)){
  primary_DF <- DF_maker(gemini_query(gemini_db, primary_IDs, sample_name), 'Primary')
  primary_IDs <- ID_to_DF(primary_IDs, 'Primary')
} else {
  primary_DF <- NA
  primary_IDs <- NA
}
if (length(secondary_IDs)>0){
  secondary_DF <- DF_maker(gemini_query(gemini_db, secondary_IDs, sample_name), 'Secondary')
  secondary_IDs <- ID_to_DF(primary_IDs, 'Secondary')
} else {
  secondary_DF <- NA
  secondary_IDs <- NA
}
```


```{r, include = FALSE}
# merge together
all_DF <- bind_rows(primary_DF, secondary_DF)
all_IDs <- bind_rows(primary_IDs, secondary_IDs)

#find failures to match ID/vcf_id and add
HGVS = all_IDs$HGVS
if (length(HGVS[!HGVS %in% all_DF$vcf_id]) > 0){
  for (i in HGVS[!HGVS %in% all_DF$vcf_id]){
    ref <- id_status %>% filter(ID==i) %>% .[['REF']]
    alt <- id_status %>% filter(ID==i) %>% .[['ALT']]
    query_result <- gemini_query(gemini_db, i, sample_name, ref, alt)
    if (length(query_result) != 2) {
      print(paste("ERROR: missing ", i)) 
      quit(status = 1)
    }
    priority <- all_IDs %>% filter(HGVS==i) %>% pull(Priority)
    all_DF <- bind_rows(all_DF, DF_maker(gemini_query(gemini_db, i, sample_name, ref, alt), priority) )
  }
}
```

```{r, include = F}
# Helper scripts to generate URLs for the tables
link_generator <- function(base_url, ID){
  if (is.na(ID) | is.na(base_url) | ID == 'None' | base_url == 'None'){
    url <- NA
  }
  else {
    ends <- strsplit(ID, ',')[[1]]
    url <- ''
    for (end in ends){
      url <- paste(paste0('<a href="', base_url, end, '" target="_blank">', end, '</a> '), url)
    }
    if (ID == 'None') {
      url <- NA
    }
  }
  return(as.character(url))
}
clinvar_link_generator <- function(base_url, ID, clinvar_sig){
  if (clinvar_sig == 'None' | is.na(clinvar_sig) | is.na(ID) | ID == 'None') {
    url <- NA
  }
  else {
    ends <- strsplit(ID, ',')[[1]]
    url <- ''
    for (end in ends){
      url <- paste(paste0('<a href="', base_url, end, '" target="_blank">', as.character(clinvar_sig), '</a> '), url)
    }
  }
  return(as.character(url))
}
google_link_generator <- function(var, gene){
  if (is.na(var) || is.na(gene) || var == 'None' || gene == 'None'){
    output <- NA
  }
  else {
    url = 'https://scholar.google.com/scholar?hl=en&q='
    var <- gsub('>','%3E',var)
    var <- strsplit(var, ':')[[1]][2]
    output <- paste0('<a href="', url, var, '+', gene, '" target="_blank">', 'Feeling&nbspLucky?', '</a>')
  }
  return(output)
}
```

```{r, include = FALSE}

failed_read <- function(failed) {
  tryCatch({
    out <- fread(failed, header = F)
    colnames(out) <- c('vcf_id','Priority')
    out
  }, error=function(e) {
    message('No failed variants file')
    return(NA)
    }
  )
}
failed <- failed_read(params$failed)
```


```{r, include=F}
# find gt index
gts_index <- grep('gts',colnames(all_DF))
colnames(all_DF)[gts_index] <- 'Genotype'
# prep for printing
## if failed file doesn't exist, then don't bind rows with it 
## otherwise merge data with failed hgvs
if (is.na(failed_read(params$failed))){
  all_DF_view <- all_DF %>% mutate(ID=paste(chrom,end,ref,alt,sep='-')) %>% select(ID, Genotype, vcf_id:Priority) %>% mutate_if(is.factor, as.character) %>% data.table()
} else { 
  all_DF_view <- all_DF %>% mutate(ID=paste(chrom,end,ref,alt,sep='-')) %>% select(ID, Genotype, vcf_id:Priority) %>% bind_rows(failed) %>% mutate_if(is.factor, as.character) %>% data.table()
}

# format DF
all_DF_view <- all_DF_view %>% 
  dplyr::mutate(clinvar_diseases = gsub('\\|',' ', as.character(clinvar_diseases)),
                variant_samples = gsub(',', ', ', as.character(variant_samples)),
                het_samples = gsub(',', ', ', as.character(het_samples)),
                hom_alt_samples = gsub(',', ', ', as.character(hom_alt_samples)),
                DeleteriousMark = ifelse((impact_severity=='HIGH' | grepl('pathog', clinvar_sig)) & as.numeric(max_aaf_all) < 0.1, 'Candidate', NA),
                impact_severity = factor(impact_severity, levels = c('HIGH','MED','LOW',NA))) 

all_DF_view$hgvsp = sapply(all_DF_view$hgvsp, function(x) strsplit(as.character(x),':')[[1]][2])
all_DF_view$ID = sapply(all_DF_view$ID, function(x) link_generator('http://gnomad.broadinstitute.org/variant/',as.character(x)))
all_DF_view$clinvar_sig = apply(all_DF_view, 1, function(x) clinvar_link_generator('https://www.ncbi.nlm.nih.gov/clinvar?term=',as.character(x['rs_ids']), as.character(x['clinvar_sig'])))
all_DF_view$rs_ids= sapply(all_DF_view$rs_ids, function(x) link_generator('https://www.ncbi.nlm.nih.gov/projects/SNP/snp_ref.cgi?rs=',as.character(x)) )
all_DF_view$'Google Scholar' = apply(all_DF_view, 1, function(x) google_link_generator(as.character(x['vcf_id']), as.character((x['gene']))) )
all_DF_view$gene = sapply(all_DF_view$gene, function(x) link_generator('https://www.omim.org/search/?search=', as.character(x)) )
all_DF_view$vcf_id = sapply(all_DF_view$vcf_id, function(x) link_generator('https://www.mutalyzer.nl/name-checker?description=',as.character(x)) )
all_DF_view <- all_DF_view %>% 
  dplyr::arrange(impact_severity) %>% 
  data.table()
colnames(all_DF_view) <- gsub('_', ' ', colnames(all_DF_view))

gts_index <- grep('Genotype',colnames(all_DF_view))
max_aaf_all <- grep('max aaf all', colnames(all_DF_view))
# find cohort counts
het_index <- grep('^num het', colnames(all_DF_view))
hom_index <- grep('^num hom alt', colnames(all_DF_view))
# scholar
gs <- grep('Scholar', colnames(all_DF_view))
# select indices for different column views
all_cols <- seq(1,ncol(all_DF_view)) 
exac <- c(1, gts_index, 3, 7, grep('exac', colnames(all_DF_view), ignore.case = T))
neg_exac <- setdiff(all_cols, exac)
core <- c(1, gts_index, 3,5,7,8,9,10,15, max_aaf_all, het_index, hom_index, gs) 
neg_core <- setdiff(all_cols, core) 
```
## `r gsub('_',' ', old_sample)` {.tabset .tabset-fade}

### Primary Variants
```{r}
if (nrow(all_DF_view %>% filter(Priority == 'Primary')) == 0){
  out <- 'No primary variants'
} else {
  out <- all_DF_view %>% 
    dplyr::filter(Priority == 'Primary') %>% 
    dplyr::arrange(DeleteriousMark, `impact severity`) %>% 
    DT::datatable(width=1200, 
                  escape = FALSE,
                  rownames = F, 
                  class='compact', 
                  extensions = c('FixedColumns','Buttons'), 
                  options = list(columnDefs=list(list(targets=neg_core-1, visible = FALSE)),
                                 pageLength = 30,
                                 lengthMenu = list(c(5, 30, 100, -1), list('5', '30', '100', 'All')),
                                 dom = 'Bfrtip',
                                 buttons = list(list(extend='pageLength'),
                                                list(extend='colvis', collectionLayout='four-column'),
                                                list(extend='colvisGroup', text='Core Fields', hide=paste(neg_core-1, collapse = ','), show = paste(core-1, collapse = ',')),
                                                list(extend='colvisGroup', text='ExAC', hide=paste(neg_exac-1,collapse = ','), show = paste(exac-1, collapse = ',' )),
                                                list(extend='colvisGroup', text='Show All', show = ':hidden' )),
                                 columnDefs=list(list(targets=c(1), class="dt-left")))) %>% 
    DT::formatSignif(columns=c('max aaf all','af exac afr','af exac amr','af exac eas','af exac oth','af exac sas','an exac all','mis z','pli'),3) %>% 
    DT::formatStyle(columns = c('DeleteriousMark'), target = 'row', backgroundColor = DT::styleEqual(c(NA, 'Candidate'), c('white', '#ff6d66')))
}
out
```

### Secondary Variants
```{r}
if (nrow(all_DF_view %>% filter(Priority == 'Secondary')) == 0){
  out <- 'No secondary variants'
} else {
  out <- all_DF_view %>% 
    dplyr::filter(Priority == 'Secondary') %>% 
    dplyr::arrange(DeleteriousMark, `impact severity`) %>% 
    DT::datatable(width=1200, 
                  escape = FALSE,
                  rownames = F, 
                  class='compact', 
                  extensions = c('FixedColumns','Buttons'), 
                  options = list(columnDefs=list(list(targets=neg_core-1, visible = FALSE)),
                                 pageLength = 30,
                                 lengthMenu = list(c(5, 30, 100, -1), list('5', '30', '100', 'All')),
                                 dom = 'Bfrtip',
                                 buttons = list(list(extend='pageLength'),
                                                list(extend='colvis', collectionLayout='four-column'),
                                                list(extend='colvisGroup', text='Core Fields', hide=paste(neg_core-1, collapse = ','), show = paste(core-1, collapse = ',')),
                                                list(extend='colvisGroup', text='ExAC', hide=paste(neg_exac-1,collapse = ','), show = paste(exac-1, collapse = ',' )),
                                                list(extend='colvisGroup', text='Show All', show = ':hidden' )),
                                 columnDefs=list(list(targets=c(1), class="dt-left")))) %>% 
    DT::formatSignif(columns=c('max aaf all','af exac afr','af exac amr','af exac eas','af exac oth','af exac sas','an exac all','mis z','pli'),3) %>% 
    DT::formatStyle(columns = c('DeleteriousMark'), target = 'row', backgroundColor = DT::styleEqual(c(NA, 'Candidate'), c('white', '#ff6d66')))
}
out
```

### Info

This report is auto-generated with scripts written by David McGaughey. This report was created on `r strftime(Sys.time(), format = "%B %d, %Y")`

**FAQ**:

- What does it mean if a variant is marked in red?
  - This means that the variant has an impact severity of ['HIGH'](http://gemini.readthedocs.io/en/latest/content/database_schema.html#details-of-the-impact-and-impact-severity-columns) **or** is marked as pathogenic in ClinVar **and** has a population allele frequency less than 10%. 

**WARNINGS/LIMITATIONS** 

This report *may* have mistakes and omissions including, but not limited to:

- Incorrect HGVS conversion to genomic coordinates which can result in the incorrect genotype for a variant
- Dropping of variants from the original .xlsx file
- Only variants from the .xlsx can be included - CNV/etc. analyses will not be reported. 
- Heterygous (num het) and Homozygous Alternative (num hom alt) allele counts **do not have a denominator** as we have no way which samples in the OGVFB cohort have the reference genotype

The definitive source of the genotype calls and analysis are the original.xlsx file and MVL/John Chiang clinical report, respectively



